fix(file): better uploaded file handling and thumbnail serving

mrclay opened the pull request:
Uploaded files now always have a new name with the client-given extension, even when replacing a previous upload.

Thumbnail JPEG files now have the `.jpg` extension and are served with the correct Content-Type. If a plugin happened to have created their own non-JPEG thumbs, they'd now be served with the correct MIME (instead of the original file's MIME type).

When an existing file is re-uploaded by a different user, the thumbnails are now created in the correct (owner's) directory.

If the file is recognized as an image, but thumbnails can't be created, we no longer allow thumbs from the previous file to be re-used.

Fixes #9612
Fixes #9267
Fixes #6677

mrclay committed:
fix(file): better uploaded file handling and thumbnail serving

Uploaded files now always have a new name with the client-given extension,
even when replacing a previous upload.

Thumbnail JPEG files now have the `.jpg` extension and are served with the
correct Content-Type. If a plugin happened to have created their own non-JPEG
thumbs, they'd now be served with the correct MIME (instead of the original
file's MIME type).

When an existing file is re-uploaded by a different user, the thumbnails are
now created in the correct (owner's) directory.

If the file is recognized as an image, but thumbnails can't be created, we
no longer allow thumbs from the previous file to be re-used.

Fixes #9612
Fixes #9267
Fixes #6677

mrclay commented on mod/file/actions/file/upload.php:
```diff
@@ -108,61 +104,76 @@
+	$thumb = new ElggFile();
+	$thumb->owner_guid = $file->owner_guid;
+
+	$sizes = [
```
@hypeJunction thx :)

hypeJunction replied:
Welcome. One of the most annoying pieces of code in Elgg. Copy pasted way too many times.

hypeJunction commented:
LGTM.
I will rebase the other PR to avoid merge conflicts once this is in.

mrclay commented:
@juho-jaakkola are we OK with this much refactoring in 1.12? I figure it's worth it to fix 3 bugs.

hypeJunction commented:
I don't think many will notice on 1.12. There is a dozen plugins all replacing the upload action.

mrclay commented:
Manual testing: 
- uploaded new PNG image: thumbs are now .jpg
- replaced image with JPG image: file has new name
- replaced with XCF image (gimp): thumbs now removed.

mrclay commented on mod/file/actions/file/upload.php:
```diff
@@ -108,61 +104,76 @@
-		
-		$thumbnail = get_resized_image_from_existing_file($file->getFilenameOnFilestore(), 60, 60, true);
-		if ($thumbnail) {
-			$thumb = new ElggFile();
```
We didn't set the owner so we were assuming the logged in user owned it. Was this another bug?

hypeJunction replied:
:)

hypeJunction commented on mod/file/actions/file/upload.php:
```diff
@@ -108,61 +104,76 @@
-		// if it is not an image, we do not need thumbnails
-		unset($file->icontime);
-		
-		$thumb = new ElggFile();
```
This is another one

hypeJunction commented:
I am merging.
